File: plugin/src/tools/cluster.ts
/**
 * Cluster Tools
 *
 * Tools for cluster and node management operations.
 */

import { Type } from "@sinclair/typebox";
import type { ClawbernnetesTool, ToolResult } from "../types.js";
import { getClient } from "../client.js";

function jsonResult(data: unknown): ToolResult {
  return {
    type: "json",
    content: JSON.stringify(data, null, 2),
  };
}

function errorResult(message: string): ToolResult {
  return {
    type: "error",
    content: message,
  };
}

// ─────────────────────────────────────────────────────────────
// cluster_status
// ─────────────────────────────────────────────────────────────

const ClusterStatusSchema = Type.Object({
  cluster: Type.Optional(Type.String({ description: "Cluster name (uses default if omitted)" })),
});

export function createClusterStatusTool(): ClawbernnetesTool {
  return {
    name: "cluster_status",
    label: "Clawbernetes",
    description:
      "Get the current status of the Clawbernetes cluster including node count, GPU availability, and workload summary.",
    parameters: ClusterStatusSchema,
    execute: async (_id, _args, context) => {
      try {
        const client = getClient(context?.config);
        const status = await client.getClusterStatus();
        return jsonResult({
          cluster: status.name,
          healthy: status.healthy,
          nodes: status.nodes,
          gpus: status.gpus,
          workloads: status.workloads,
          summary: `Cluster ${status.name}: ${status.nodes.ready}/${status.nodes.total} nodes ready, ${status.gpus.available}/${status.gpus.total} GPUs available, ${status.workloads.running} workloads running`,
        });
      } catch (err) {
        return errorResult(`Failed to get cluster status: ${err}`);
      }
    },
  };
}

// ─────────────────────────────────────────────────────────────
// node_list
// ─────────────────────────────────────────────────────────────

const NodeListSchema = Type.Object({
  status: Type.Optional(
    Type.String({
      description: "Filter by status: ready, not_ready, cordoned, draining, offline",
    })
  ),
  labels: Type.Optional(
    Type.Record(Type.String(), Type.String(), {
      description: "Filter by labels (key=value pairs)",
    })
  ),
  gpuModel: Type.Optional(Type.String({ description: "Filter by GPU model (e.g., H100, A100)" })),
});

export function createNodeListTool(): ClawbernnetesTool {
  return {
    name: "node_list",
    label: "Clawbernetes",
    description:
      "List all nodes in the cluster with optional filtering by status, labels, or GPU model.",
    parameters: NodeListSchema,
    execute: async (_id, args, context) => {
      try {
        const client = getClient(context?.config);
        const params = args as { status?: string; labels?: Record<string, string>; gpuModel?: string };
        let nodes = await client.listNodes({
          status: params.status,
          labels: params.labels,
        });

        // Filter by GPU model if specified
        if (params.gpuModel) {
          const model = params.gpuModel.toLowerCase();
          nodes = nodes.filter((n) =>
            n.gpus.some((g) => g.model.toLowerCase().includes(model))
          );
        }

        return jsonResult({
          count: nodes.length,
          nodes: nodes.map((n) => ({
            id: n.id,
            name: n.name,
            status: n.status,
            gpus: n.gpus.length,
            gpuModels: [...new Set(n.gpus.map((g) => g.model))],
            region: n.region,
            lastHeartbeat: new Date(n.lastHeartbeat).toISOString(),
          })),
        });
      } catch (err) {
        return errorResult(`Failed to list nodes: ${err}`);
      }
    },
  };
}

// ─────────────────────────────────────────────────────────────
// node_get
// ─────────────────────────────────────────────────────────────

const NodeGetSchema = Type.Object({
  nodeId: Type.String({ description: "Node ID or name" }),
});

export function createNodeGetTool(): ClawbernnetesTool {
  return {
    name: "node_get",
    label: "Clawbernetes",
    description: "Get detailed information about a specific node including GPU status and workloads.",
    parameters: NodeGetSchema,
    execute: async (_id, args, context) => {
      try {
        const client = getClient(context?.config);
        const { nodeId } = args as { nodeId: string };
        const node = await client.getNode(nodeId);

        if (!node) {
          return errorResult(`Node not found: ${nodeId}`);
        }

        return jsonResult({
          id: node.id,
          name: node.name,
          status: node.status,
          platform: node.platform,
          region: node.region,
          zone: node.zone,
          resources: {
            cpuCores: node.cpuCores,
            memoryMb: node.memoryMb,
            gpus: node.gpus.map((g) => ({
              id: g.id,
              model: g.model,
              vendor: g.vendor,
              memoryMb: g.memoryMb,
              utilization: g.utilizationPercent,
              temperature: g.temperatureCelsius,
            })),
          },
          labels: node.labels,
          connectedAt: new Date(node.connectedAt).toISOString(),
          lastHeartbeat: new Date(node.lastHeartbeat).toISOString(),
        });
      } catch (err) {
        return errorResult(`Failed to get node: ${err}`);
      }
    },
  };
}

// ─────────────────────────────────────────────────────────────
// node_drain
// ─────────────────────────────────────────────────────────────

const NodeDrainSchema = Type.Object({
  nodeId: Type.String({ description: "Node ID or name to drain" }),
  gracePeriodSeconds: Type.Optional(
    Type.Number({ description: "Grace period for workload migration (default: 300)" })
  ),
  force: Type.Optional(
    Type.Boolean({ description: "Force drain even if workloads cannot be migrated" })
  ),
});

export function createNodeDrainTool(): ClawbernnetesTool {
  return {
    name: "node_drain",
    label: "Clawbernetes",
    description:
      "Drain a node by migrating all workloads to other nodes. Use before maintenance or decommissioning.",
    parameters: NodeDrainSchema,
    execute: async (_id, args, context) => {
      try {
        const client = getClient(context?.config);
        const { nodeId, gracePeriodSeconds, force } = args as {
          nodeId: string;
          gracePeriodSeconds?: number;
          force?: boolean;
        };

        const result = await client.drainNode(nodeId, { gracePeriodSeconds, force });

        if (result.success) {
          return jsonResult({
            success: true,
            nodeId,
            migratedWorkloads: result.migratedWorkloads,
            message: `Successfully drained node ${nodeId}. Migrated ${result.migratedWorkloads} workloads.`,
          });
        } else {
          return errorResult(`Failed to drain node ${nodeId}`);
        }
      } catch (err) {
        return errorResult(`Failed to drain node: ${err}`);
      }
    },
  };
}

// ─────────────────────────────────────────────────────────────
// node_cordon / node_uncordon
// ─────────────────────────────────────────────────────────────

const NodeCordonSchema = Type.Object({
  nodeId: Type.String({ description: "Node ID or name" }),
});

export function createNodeCordonTool(): ClawbernnetesTool {
  return {
    name: "node_cordon",
    label: "Clawbernetes",
    description:
      "Cordon a node to prevent new workloads from being scheduled on it. Existing workloads continue running.",
    parameters: NodeCordonSchema,
    execute: async (_id, args, context) => {
      try {
        const client = getClient(context?.config);
        const { nodeId } = args as { nodeId: string };
        const result = await client.cordonNode(nodeId);

        if (result.success) {
          return jsonResult({
            success: true,
            nodeId,
            message: `Node ${nodeId} cordoned. No new workloads will be scheduled.`,
          });
        } else {
          return errorResult(`Failed to cordon node ${nodeId}`);
        }
      } catch (err) {
        return errorResult(`Failed to cordon node: ${err}`);
      }
    },
  };
}

export function createNodeUncordonTool(): ClawbernnetesTool {
  return {
    name: "node_uncordon",
    label: "Clawbernetes",
    description: "Uncordon a node to allow new workloads to be scheduled on it again.",
    parameters: NodeCordonSchema,
    execute: async (_id, args, context) => {
      try {
        const client = getClient(context?.config);
        const { nodeId } = args as { nodeId: string };
        const result = await client.uncordonNode(nodeId);

        if (result.success) {
          return jsonResult({
            success: true,
            nodeId,
            message: `Node ${nodeId} uncordoned. Workloads can now be scheduled.`,
          });
        } else {
          return errorResult(`Failed to uncordon node ${nodeId}`);
        }
      } catch (err) {
        return errorResult(`Failed to uncordon node: ${err}`);
      }
    },
  };
}

// ─────────────────────────────────────────────────────────────
// Export all cluster tools
// ─────────────────────────────────────────────────────────────

export function createClusterTools(): ClawbernnetesTool[] {
  return [
    createClusterStatusTool(),
    createNodeListTool(),
    createNodeGetTool(),
    createNodeDrainTool(),
    createNodeCordonTool(),
    createNodeUncordonTool(),
  ];
}

File: crates/claw-bridge/src/handlers/cluster.rs
//! Cluster management handlers
//!
//! These handlers integrate with claw-discovery and claw-compute to provide
//! cluster-level operations.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

use crate::error::{BridgeError, BridgeResult};
use crate::handlers::{parse_params, to_json};

// ─────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct ClusterStatus {
    pub name: String,
    pub healthy: bool,
    pub nodes: NodeCounts,
    pub gpus: GpuCounts,
    pub workloads: WorkloadCounts,
}

#[derive(Debug, Clone, Serialize)]
pub struct NodeCounts {
    pub total: u32,
    pub ready: u32,
    pub not_ready: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct GpuCounts {
    pub total: u32,
    pub available: u32,
    pub allocated: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct WorkloadCounts {
    pub running: u32,
    pub pending: u32,
    pub failed: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct NodeInfo {
    pub id: String,
    pub name: String,
    pub status: String,
    pub labels: HashMap<String, String>,
    pub gpus: Vec<GpuInfo>,
    pub cpu_cores: u32,
    pub memory_mb: u64,
    pub platform: String,
    pub region: Option<String>,
    pub zone: Option<String>,
    pub connected_at: u64,
    pub last_heartbeat: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct GpuInfo {
    pub id: String,
    pub model: String,
    pub vendor: String,
    pub memory_mb: u64,
    pub compute_capability: Option<String>,
    pub utilization_percent: Option<f32>,
    pub temperature_celsius: Option<f32>,
}

// ─────────────────────────────────────────────────────────────
// Handlers
// ─────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct StatusParams {
    pub cluster: Option<String>,
}

/// Get cluster status
pub async fn status(params: Value) -> BridgeResult<Value> {
    let _params: StatusParams = parse_params(params)?;

    // TODO: Aggregate from actual crates
    // For now, return mock data that matches the expected interface
    let status = ClusterStatus {
        name: "default".to_string(),
        healthy: true,
        nodes: NodeCounts {
            total: 0,
            ready: 0,
            not_ready: 0,
        },
        gpus: GpuCounts {
            total: 0,
            available: 0,
            allocated: 0,
        },
        workloads: WorkloadCounts {
            running: 0,
            pending: 0,
            failed: 0,
        },
    };

    to_json(status)
}

#[derive(Debug, Deserialize)]
pub struct NodeListParams {
    pub status: Option<String>,
    pub labels: Option<HashMap<String, String>>,
    pub gpu_model: Option<String>,
}

/// List nodes in the cluster
pub async fn node_list(params: Value) -> BridgeResult<Value> {
    let _params: NodeListParams = parse_params(params)?;

    // TODO: Get from claw-discovery service registry
    let nodes: Vec<NodeInfo> = vec![];

    to_json(nodes)
}

#[derive(Debug, Deserialize)]
pub struct NodeGetParams {
    pub node_id: String,
}

/// Get detailed node information
pub async fn node_get(params: Value) -> BridgeResult<Value> {
    let params: NodeGetParams = parse_params(params)?;

    // TODO: Look up in claw-discovery
    Err(BridgeError::NotFound(format!(
        "node not found: {}",
        params.node_id
    )))
}

#[derive(Debug, Deserialize)]
pub struct NodeDrainParams {
    pub node_id: String,
    pub grace_period_seconds: Option<u32>,
    pub force: Option<bool>,
}

#[derive(Debug, Serialize)]
pub struct NodeDrainResult {
    pub success: bool,
    pub migrated_workloads: u32,
}

/// Drain a node (migrate all workloads off)
pub async fn node_drain(params: Value) -> BridgeResult<Value> {
    let params: NodeDrainParams = parse_params(params)?;

    // TODO: Implement via claw-preemption + workload migration
    tracing::info!(
        node_id = %params.node_id,
        force = params.force.unwrap_or(false),
        "draining node"
    );

    to_json(NodeDrainResult {
        success: true,
        migrated_workloads: 0,
    })
}

#[derive(Debug, Deserialize)]
pub struct NodeCordonParams {
    pub node_id: String,
}

#[derive(Debug, Serialize)]
pub struct NodeCordonResult {
    pub success: bool,
}

/// Cordon a node (prevent new workload scheduling)
pub async fn node_cordon(params: Value) -> BridgeResult<Value> {
    let params: NodeCordonParams = parse_params(params)?;

    // TODO: Update node state in discovery
    tracing::info!(node_id = %params.node_id, "cordoning node");

    to_json(NodeCordonResult { success: true })
}

/// Uncordon a node (allow new workload scheduling)
pub async fn node_uncordon(params: Value) -> BridgeResult<Value> {
    let params: NodeCordonParams = parse_params(params)?;

    // TODO: Update node state in discovery
    tracing::info!(node_id = %params.node_id, "uncordoning node");

    to_json(NodeCordonResult { success: true })
}
